src/odemis/driver/tfsbc.py

```diff
@@ -0,0 +1,357 @@
+
+You should have received a copy of the GNU General Public License along with
+Odemis. If not, see http://www.gnu.org/licenses/.
+"""
+from __future__ import division
+
+import logging
+
+import numpy
+import serial
+import serial.tools.list_ports
+from pymodbus.client.sync import ModbusSerialClient
+
+from odemis import model
+from odemis.model import HwError
+
+# Parameters for connection
+BAUDRATE = 230400
+TIMEOUT = 1
+BYTESIZE = 8
+PARITY = serial.PARITY_NONE
+STOPBITS = 1
+
+# Modbus level addresses
+SLAVE_UNIT = 2
+
+# Modbus registers
+BEAMDEFL_LX = 0  # lower x beam deflection element control
+BEAMDEFL_LY = 1  # lower y beam deflection element control
+BEAMDEFL_UX = 2  # upper x beam deflection element control
+BEAMDEFL_UY = 3  # upper y beam deflection element control
+
+# Transformation constants
+DCREVERSCOEF = -1
+DCROTUPPXX = 1
+DCROTUPPYY = 1
+DCROTLOWXX = -1
+DCROTLOWYY = 1
+
+# Conversion ranges
+C_MIN_RAW_SHIFT = 0
+C_MAX_RAW_SHIFT = 0xFFFF
+C_MIN_DBL_SHIFT = -42.2e-3
+C_MAX_DBL_SHIFT = 42.2e-3
+
+
+def current_to_raw(current):
+    """
+    Helper function for coordinate transform (from Thermofischer example code).
+    :param current: (float)
+    """
+    k = (C_MAX_RAW_SHIFT - C_MIN_RAW_SHIFT) / (C_MAX_DBL_SHIFT - C_MIN_DBL_SHIFT)
+    return int((current - C_MIN_DBL_SHIFT) * k + C_MIN_RAW_SHIFT + 0.5)
+
+
+def raw_to_current(raw):
+    """
+    Inverse of current_to_raw.
+    :param raw: (int)
+    """
+    k = (C_MAX_RAW_SHIFT - C_MIN_RAW_SHIFT) / (C_MAX_DBL_SHIFT - C_MIN_DBL_SHIFT)
+    return (raw - 0.5 - C_MIN_RAW_SHIFT) / k + C_MIN_DBL_SHIFT
+
+
+def transform_coordinates(value, xlower, ylower, xupper, yupper):
+    """
+    Transform x, y coordinates to register values of beamshift hardware.
+    :param value: (float, float) x, y value in the source coordinate system
+    :param xlower: (float, float) xlower metadata
+    :param ylower: (float, float) ylower metadata
+    :param xupper: (float, float) xupper metadata
+    :param yupper: (float, float) yupper metadata
+    :return (int, int, int, int): register values: x lower, y lower, x upper, y upper
+    """
+    value = (value[0] * 1e6, value[1] * 1e6)  # value in µm
+
+    # This transformation was provided as example code from Thermofischer (the variable names are slightly modified
+    # to fit the coding style of this driver, the rest of the calculation is identical).
+    dc_xupper = value[0] * xupper[0] + value[1] * yupper[0]
+    dc_xlower = value[0] * xlower[0] + value[1] * ylower[0]
+    dc_yupper = value[0] * xupper[1] + value[1] * yupper[1]
+    dc_ylower = value[0] * xlower[1] + value[1] * ylower[1]
+
+    currUX = DCREVERSCOEF * DCROTUPPXX * dc_xupper
+    currLX = DCREVERSCOEF * DCROTLOWXX * dc_xlower
+    currUY = DCREVERSCOEF * DCROTUPPYY * dc_yupper
+    currLY = DCREVERSCOEF * DCROTLOWYY * dc_ylower
+
+    for current in [currLX, currLY, currUX, currUY]:
+        if current < C_MIN_DBL_SHIFT or current > C_MAX_DBL_SHIFT:
```

Comment:
You can write slightly more clearly:
`if not C_MIN_DBL_SHIFT <= current <= C_MAX_DBL_SHIFT:`